hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py

```diff
@@ -19,19 +20,19 @@
 
 class Float(Shrinker):
     def setup(self):
-        self.NAN = math.nan
         self.debugging_enabled = True
 
-    def make_immutable(self, f):
-        f = float(f)
+    def make_canonical(self, f):
         if math.isnan(f):
-            # Always use the same NAN so it works properly in self.seen
-            f = self.NAN
+            # Distinguish different NaN bit patterns, while making each equal to itself.
+            # Returning bytes instead of integer (float_to_int) avoids accidental
+            # equality with valid large floats.
+            return struct.pack("d", f)
```

Comment:
I'm surprised we can't use `float_to_int` for a unique key here - did you find two floats with the same `float_to_int` value? that would be bad news for places where we rely on `float_to_int` for equality (see `choice_equal`)

Reply:
We could use `float_to_int` unconditionally, but here we're only transforming the nans (for performance), so the int could be compared to a plain float.

Reply:
gotcha. I'd prefer `("nan", float_to_int(f))` or `int_to_bytes(float_to_int(f))` so we don't forget to update this when supporting different float widths (https://github.com/HypothesisWorks/hypothesis/issues/3959).

Reply:
eh, I think that's far enough away and will be enough of a rewrite that I'm not worried, and I like the simplicity of this approach in the meantime.

Reply:
I think the suggested tuple is easier to understand the purpose of though, so I'll go with that.